### PR TITLE
first commit

### PR DESCRIPTION
**Pull request by AyeletKeren**

### CODE REVIEW

File: tests/blackbox/conftest.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            f'RedisConnectionString={self.redis_host}',
+            'RedisPassword=',
+            f'redis_port={self.redis_port}',
+            f'redis_host={self.redis_host}',
+
+            f'rabbitmq_port={self.rabbitmq_port}',
+            f'rabbitmq_host={self.rabbitmq_host}',
+            f'rabbitmq_exchange={self.rabbitmq_exchange}',
+            f'rabbitmq_queue_name={self.rabbitmq_queue_name}',
+            f'RABBIT_URL=amqp://guest:guest@{self.rabbitmq_host}:{self.rabbitmq_port}/guest',
```
Comment (bentheiii) [inline tests/blackbox/conftest.py:43]: why list the host and port twice?

File: tests/blackbox/image/conftest.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            connect(network, blob_storage):
+        env_vars.redis_host = redis_alias[0]
+        env_vars.rabbitmq_host = rabbit_alias[0]
+        env_vars.connection_string = blob_storage.connection_string
+
+        env = env_vars.as_dotenv()
+
+        container = docker_client.containers.create(
+            'biocatchtest/ayelet_initmanager:testing',
+            ports={SERVICE_PORT: SERVICE_PORT},
```
Comment (bentheiii) [inline tests/blackbox/image/conftest.py:48]: don't autobind

```suggestion
            ports={SERVICE_PORT: 0},
```

File: tests/blackbox/image/conftest.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                                        tag='biocatchtest/ayelet_initmanager:testing',
+                                        rm=True)
+    for msg_b in build_log:
+        msgs = str(msg_b, 'utf-8').splitlines()
+        for msg in msgs:
+            s = json.loads(msg).get('stream')
+            if s:
+                print(s, end='', flush=True, file=sys.stderr)
+            else:
+                print(msg, file=sys.stderr)
```
Comment (bentheiii) [inline tests/blackbox/image/conftest.py:34]: U2U but you can use yellowbox for this

File: tests/blackbox/image/conftest.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+            def pipe():
+                for line_b in log_stream:
+                    line = str(line_b, 'utf-8').strip()
+                    print(line, file=sys.stderr)
+
+            pipe_thread = Thread(target=pipe)
+            pipe_thread.start()
+
+            yield container
```
Comment (bentheiii) [inline tests/blackbox/image/conftest.py:64]: retry on the health check instead of sleeping

File: tests/blackbox/image/test_image.py [hunk 5]
```diff
@@ -0,0 +1,9 @@
+from time import sleep
+
+import requests
+
+
+def test_readiness(im, base_url):
+    sleep(5)
```
Comment (bentheiii) [inline tests/blackbox/image/test_image.py:7]: don't sleep, retry on the container health

File: tests/blackbox/image/conftest.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    return f'http://localhost:{SERVICE_PORT}/api'
+
+
+@fixture
+def v1_url(base_url):
+    return base_url + '/v1'
+
+
+@fixture
+def im(docker_client, env_name, env_vars, redis, rabbitmq, blob_storage):
```
Comment (bentheiii) [inline tests/blackbox/image/conftest.py:22]: it's called im but it returns a container?

File: tests/unittest/test_blob.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
+from unittest.mock import Mock
+
+import pytest
+from _pytest.monkeypatch import MonkeyPatch
```
Comment (bentheiii) [inline tests/unittest/test_blob.py:4]: never import _pytest

File: tests/unittest/test_blob.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+@pytest.fixture
+def blob_mock(monkeypatch: MonkeyPatch) -> BlobServiceClient:
+    monkeypatch.setenv('connection_string', connection_string)
+
+    def from_connection_string(connection_string):
+        return None
+
+    monkeypatch.setattr(BlobServiceClient, 'from_connection_string',
+                        Mock(wraps=from_connection_string))
```
Comment (bentheiii) [inline tests/unittest/test_blob.py:23]:
```suggestion
                        MagicMock())
```

File: tests/unittest/test_init_messages_processor.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    return AsyncMock()
+
+
+@pytest.mark.asyncio
+async def test_init_message_handling(init_message_1, blob_mock, redis_mock):
+    with patch('utils.blob.download_user_data', return_value=PROFILES) as blob_download_mock:
+        with patch('utils.redis.hsetnx', return_value=None) as redis_hsetnx_mock:
+            await handle_rmq_message(init_message_1, CONTAINER, PROF_PREF, DS_PROF_PREF, blob_mock, redis_mock)
+            blob_download_mock.assert_called_with(_uid, CONTAINER, PROF_PREF, DS_PROF_PREF, blob_mock)
+            [redis_hsetnx_mock.assert_called_with(_uid, field, PROFILES[field], redis_mock) for field in PROFILES]
```
Comment (bentheiii) [inline tests/unittest/test_init_messages_processor.py:38]: never do this, assert one by one

File: tests/unittest/test_redis.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
+from unittest.mock import Mock
+
+import aioredis
+import pytest
+from _pytest.monkeypatch import MonkeyPatch
```
Comment (bentheiii) [inline tests/unittest/test_redis.py:5]: _pytest

File: utils/blob.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
+import asyncio
```
Comment (bentheiii) [inline utils/blob.py:1]: this entire file should be either deleted or moved into app

File: utils/rabbitmq.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
+import logging
+from typing import Any, Callable
+
+import aio_pika
+import envolved
+from aio_pika import IncomingMessage
+from aio_pika.abc import AbstractRobustConnection
+
+logger = logging.getLogger(('biocatch.' + __name__))
```
Comment (bentheiii) [inline utils/rabbitmq.py:9]:
```suggestion
logger = logging.getLogger('biocatch.' + __name__)
```

File: utils/rabbitmq.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
+import logging
```
Comment (bentheiii) [inline utils/rabbitmq.py:1]: move to app

File: utils/redis.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
+import logging
+import zlib
+
+import aioredis
+import envolved
+import ormsgpack
+from aioredis import Redis
+
+logger = logging.getLogger('biocatch.' + __name__)
+TIMEOUT = 5000  # ms
```
Comment (bentheiii) [inline utils/redis.py:10]: better name

File: utils/redis.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
+import logging
```
Comment (bentheiii) [inline utils/redis.py:1]: move to app

File: utils/redis.py [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+        port = redis_port.get()
+        return await aioredis.create_redis_pool(
+            address=(host, port), timeout=TIMEOUT)
+    except Exception:
+        logger.exception("Redis connection failed")
+        raise
+
+
+async def hsetnx(key: str, field: str, value: bytes, redis: Redis) -> None:
+    await redis.hsetnx(key, field, value)
```
Comment (bentheiii) [inline utils/redis.py:29]: why?

File: utils/redis.py [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+
+
+async def hget(key: str, field: str, redis: Redis) -> bytes:
+    profile = await redis.hget(key, field)
+    decompressed_profile = zlib.decompress(profile)
+    deserialized_profile = ormsgpack.unpackb(decompressed_profile)
+    return deserialized_profile
```
Comment (bentheiii) [inline utils/redis.py:36]: do this manually

**Review by bentheiii — CHANGES_REQUESTED** [review]

File: app/initmanager/init_message_consumer.py [hunk 18]
```diff
@@ -7,22 +7,26 @@
...
-        uid = list(parsed_message.values())[0]
+        uid = parsed_message[MESSAGE_KEY]
     except Exception:
-        logger.exception(f'failed to parse incoming message: {incoming_message}')
+        logger.exception(f':{PARSING_FAILURE_MESSAGE} {incoming_message}')
     try:
         blob_profiles = await blob.download_user_data(uid, container, profiles_pref, ds_profiles_pref,
                                                       blob_service_client)
-        await gather(*[redis.hsetnx(uid, key, blob_profiles[key], aioredis) for key in blob_profiles])
+        await gather(*[aioredis.hset(uid, key, blob_profiles[key]) for key in blob_profiles])
```
Comment (bentheiii) [inline app/initmanager/init_message_consumer.py:27]: hset can set multiple keys at once

File: app/main.py [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+import envolved
+import uvicorn
+from aio_pika import IncomingMessage
+from fastapi import FastAPI
+
+from app.initmanager.init_message_consumer import handle_rmq_message
+from utils import rabbitmq, redis
+from utils.blob import create_blob_client
+
+logger = logging.getLogger('biocatch.' + __name__)
```
[the remark this answers is not in the document]
Comment (bentheiii) [reply app/main.py:18]: still this

File: app/main.py [hunk 20]
```diff
@@ -1,17 +1,36 @@
...
-from utils import rabbitmq, redis
-from utils.blob import create_blob_client
+from app.utils import rabbitmq
+from app.utils.blob import create_blob_client
+
+INIT_MESSAGE_PROCESS_FAILURE_MESSAGE = 'exception when processing init message on rabbit mq'
+PREPARE_INIT_MANAGER_FAILURE_MESSAGE = 'exception when preparing init manager environment'
 
 logger = logging.getLogger('biocatch.' + __name__)
+TIMEOUT = 10000  # ms
```
Comment (bentheiii) [inline app/main.py:19]: better

File: app/main.py [hunk 21]
```diff
@@ -0,0 +1,83 @@
+import logging
+from asyncio import gather
+from typing import Tuple
+
+import envolved
```
[the remark this answers is not in the document]
Comment (bentheiii) [reply app/main.py:7]: still this

File: app/main.py [hunk 22]
```diff
@@ -1,17 +1,36 @@
...
 
 logger = logging.getLogger('biocatch.' + __name__)
+TIMEOUT = 10000  # ms
+
+
+async def create_connection_pool() -> Redis:
+    redis_host: envolved.EnvVar[str] = \
+        envolved.env_var('redis_host', type=str)
+    redis_port: envolved.EnvVar[str] = \
+        envolved.env_var('redis_port', type=str)
```
Comment (bentheiii) [inline app/main.py:26]: why the line breaks?

Comment (bentheiii) [inline app/main.py:26]: shouldn't this be an int?

File: app/main.py [hunk 23]
```diff
@@ -65,11 +83,18 @@ async def shutdown_event() -> None:
...
+    profile, ds_profile = await gather(get(uid, get_key(app.profiles_pref, uid), app.redis),
+                                       get(uid, get_key(app.ds_profiles_pref, uid), app.redis))
     return profile, ds_profile
 
 
+async def get(key: str, field: str, redis: Redis) -> bytes:
+    profile = await redis.hget(key, field)
+    decompressed_profile = zlib.decompress(profile)
+    deserialized_profile = ormsgpack.unpackb(decompressed_profile)
+    return deserialized_profile
```
Comment (bentheiii) [inline app/main.py:92]: why not do hgetall?

File: tests/unittest/test_blob.py (hunk 7 above)
Comment (bentheiii) [reply tests/unittest/test_blob.py:4]: still

File: app/main.py (hunk 22 above)
Comment (AyeletKeren) [reply app/main.py:26]: I remember you told me all yellowbox env vars should be string (?..) that's why I used strings

File: app/main.py [hunk 24]
```diff
@@ -1,71 +1,54 @@
...
 
 from app.initmanager.init_message_consumer import handle_rmq_message
 from app.utils import rabbitmq
 from app.utils.blob import create_blob_client
 
-INIT_MESSAGE_PROCESS_FAILURE_MESSAGE = 'exception when processing init message on rabbit mq'
-PREPARE_INIT_MANAGER_FAILURE_MESSAGE = 'exception when preparing init manager environment'
-
-logger = logging.getLogger('biocatch.' + __name__)
+logger = logging.getLogger('biocatch.ayelet_init_manager')
```
Comment (bentheiii) [inline app/main.py:14]:
```suggestion
logger = logging.getLogger('biocatch.' + __name__)
```

File: app/initmanager/init_message_consumer.py [hunk 25]
```diff
@@ -1,32 +1,27 @@
...
 
 logger = logging.getLogger('biocatch.' + __name__)
-MESSAGE_KEY = "uid"
-PARSING_FAILURE_MESSAGE = "failed to parse incoming message"
-LOAD_FAILURE_MESSAGE = "Failed to load profiles"
 
 
 async def handle_rmq_message(incoming_message: IncomingMessage,  container: str, profiles_pref: str,
-                             ds_profiles_pref: str, blob_service_client: BlobServiceClient, aioredis: Redis)\
+                             ds_profiles_pref: str, blob_service_client: BlobServiceClient, redisinst)\
```
Comment (bentheiii) [inline app/initmanager/init_message_consumer.py:14]: type hint on `redisinst`?

File: app/main.py [hunk 26]
```diff
@@ -1,71 +1,54 @@
...
 
 
-async def create_connection_pool() -> Redis:
-    redis_host: envolved.EnvVar[str] = \
-        envolved.env_var('redis_host', type=str)
-    redis_port: envolved.EnvVar[str] = \
-        envolved.env_var('redis_port', type=str)
+async def create_connection_pool():
+    redis_host: EnvVar[str] = env_var('redis_host', type=str)
+    redis_port: EnvVar[str] = env_var('redis_port', type=str)
```
Comment (bentheiii) [inline app/main.py:20]: FYI this will work, but there's no reason not to make this an int

File: app/main.py [hunk 27]
```diff
@@ -82,17 +65,16 @@ async def shutdown_event() -> None:
...
-
-
-async def get(key: str, field: str, redis: Redis) -> bytes:
-    profile = await redis.hget(key, field)
-    decompressed_profile = zlib.decompress(profile)
+async def get_data(uid: str):
+    profiles = await app.redis.hgetall(uid)
+    if (profiles == {}):
+        return 'user data missing'
+    profiles_decoded = {key.decode(): profiles.get(key) for key in profiles.keys()}
```
Comment (bentheiii) [inline app/main.py:72]:
```suggestion
    profiles_decoded = {key.decode(): value for key, value in profiles.items()}
```

Comment (bentheiii) [reply app/main.py:72]: also, these profiles aren't decoded

File: pyproject.toml [hunk 28]
```diff
@@ -6,31 +6,28 @@ authors = ["Your Name <you@example.com>"]
...
-uvicorn = "^0.17.6"
-msgpack = "^1.0.3"
 rabbitmq = "^0.2.0"
 orjson = "^3.6.8"
 ormsgpack = "^1.2.2"
+
+
+[tool.poetry.dev-dependencies]
 azure-storage-blob = "~12.11.0"
 envolved = "^0.5.0"
```
Comment (bentheiii) [inline pyproject.toml:19]: envolved isn't dev

File: tests/blackbox/app/test_blackbox.py [hunk 29]
```diff
@@ -1,88 +1,75 @@
...
-        rabbitmq_queue_name: envolved.EnvVar[str] = \
-            envolved.env_var('rabbitmq_queue_name', type=str)
-        queue = rabbitmq_queue_name.get()
-        rabbitmq_exchange: envolved.EnvVar[str] = \
-            envolved.env_var('rabbitmq_exchange', type=str)
-        exchange = rabbitmq_exchange.get()
-        channel.basic_publish(exchange, queue, message.json())
-    except Exception:
-        raise
+    response_list = orjson.loads(resp.content.decode())
```
Comment (bentheiii) [inline tests/blackbox/app/test_blackbox.py:60]:
```suggestion
    response_list = resp.json()
```

File: tests/blackbox/app/test_blackbox.py [hunk 30]
```diff
@@ -1,88 +1,75 @@
...
+
+
+@pytest.mark.asyncio
+async def test_error_on_get_missing_data(env_vars: BlackboxEnv, initmanager_client: TestClient, rabbit_channel,
+                                         redis_client, blob_storage, storage_client):
+    send_message(env_vars, rabbit_channel, INIT_MESSAGE_2)
+    await sleep(10)
+    resp = await initmanager_client.get(f'/api/v1/get-data?uid={_uid}')
+    assert resp.ok
+    assert resp.content == b'"user data missing"'
```
Comment (bentheiii) [inline tests/blackbox/app/test_blackbox.py:71]: if it's missing why are you returning OK?
(a general rule of thumb is that each endpoint's OK response should only have one valid schema)

File: app/main.py [hunk 31]
```diff
@@ -64,17 +67,17 @@ async def shutdown_event() -> None:
...
+@app.get("/api/v1/get-data", response_model=GetDataResponse)
 async def get_data(uid: str):
     profiles = await app.redis.hgetall(uid)
-    if (profiles == {}):
-        return 'user data missing'
-    profiles_decoded = {key.decode(): profiles.get(key) for key in profiles.keys()}
-    decompressed_profile = zlib.decompress(profiles_decoded[get_key(app.profiles_pref, uid)])
-    decompressed_ds_profile = zlib.decompress(profiles_decoded[get_key(app.ds_profiles_pref, uid)])
+    if profiles == {}:
+        return Response(status_code=HTTP_500_INTERNAL_SERVER_ERROR,
```
Comment (bentheiii) [inline app/main.py:74]: you should never return a 500 yourself, 500 is an unhandled exception. in this case a 404 is good

File: app/model.py [hunk 32]
```diff
@@ -0,0 +1,7 @@
+from typing import List
+
+from pydantic import BaseModel
+
+
+class GetDataResponse(BaseModel):
+    data: List[str] = []
```
Comment (bentheiii) [inline app/model.py:7]: why the default?

**Pull request merged**